pkg/jobrunaggregator/jobrunaggregatorlib/util.go

```diff
@@ -112,15 +132,110 @@ func WaitAndGetAllFinishedJobRuns(ctx context.Context,
 			case <-time.After(10 * time.Minute):
 				continue
 			case <-ctx.Done():
-				return finishedJobRuns, unfinishedJobRuns, finishedJobRunNames, unfinishedJobRunNames, ctx.Err()
+				return ctx.Err()
 			}
 		}
 
 		break
 	}
+	return nil
+}
+
+type ClusterJobRunWaiter struct {
+	ProwJobClient      *prowjobclientset.Clientset
+	TimeToStopWaiting  time.Time
+	ProwJobMatcherFunc ProwJobMatcherFunc
+	nextLogTime        time.Time
+}
+
+func (w *ClusterJobRunWaiter) allProwJobFinished(allItems []*prowv1.ProwJob) bool {
+	uncompletedJobMap := map[string]*prowv1.ProwJob{}
+	for _, prowJob := range allItems {
+		if !w.ProwJobMatcherFunc(prowJob) {
+			continue
+		}
+		if prowJob.Status.CompletionTime != nil {
+			continue
+		}
+
+		jobRunID := prowJob.Labels[prowJobJobRunIDLabel]
+		uncompletedJobMap[jobRunID] = prowJob
+	}
+	if len(uncompletedJobMap) == 0 {
+		return true
+	}
+	logrus.Infof("Current time %+v and next lgo time: [%v]", time.Now(), w.nextLogTime)
+	if time.Now().After(w.nextLogTime) {
+		logrus.Infof("The following jobs are not complete: [%v]", strings.Join(sets.StringKeySet(uncompletedJobMap).List(), ", "))
+		w.nextLogTime = time.Now().Add(unfinishedProwJobLoggingInterval)
+	}
+	return false
+}
+
+func (w *ClusterJobRunWaiter) Wait(ctx context.Context) error {
+	if w.ProwJobClient == nil {
+		return fmt.Errorf("prowjob client is missing")
+	}
+
+	prowJobInformerFactory := prowjobinformers.NewSharedInformerFactory(w.ProwJobClient, 24*time.Hour)
+	prowJobInformer := prowJobInformerFactory.Prow().V1().ProwJobs()
+
+	// done to be sure that the informer is shown as "active" so that start activates them
+	// start informers and wait for them to sync
+	hasSynced := prowJobInformer.Informer().HasSynced
+	go prowJobInformerFactory.Start(ctx.Done())
+	if !cache.WaitForCacheSync(ctx.Done(), hasSynced) {
+		return fmt.Errorf("prowjob informer sync error")
+	}
+	timeout := time.Until(w.TimeToStopWaiting)
+	if timeout < 0 {
+		timeout = 2 * time.Second
+	}
+	logrus.Infof("Going to wait until %+v with timeout value %+v", w.TimeToStopWaiting, timeout)
+
+	// wait for up to limit until we've finished
+	err := wait.PollUntilContextTimeout(
+		ctx,
+		5*time.Minute,
+		timeout,
+		true,
+		func(ctx context.Context) (bool, error) {
+			allItems, err := prowJobInformer.Lister().List(labels.Everything())
```

Comment:
What was your test process like? I'm curious if our aggregated pods have permissions to list all ProwJobs, and if not, how do we get it.

Reply:
This will be the next thing to figure out. In my test environment, just using my own token will be sufficient. I might need service account route for this.